File: Smallest Positive missing number - GFG/smallest-positive-missing-number.py
#User function Template for python3

class Solution:
    
    #Function to find the smallest positive number missing from the array.
    def missingNumber(self,nums,n):
        #Your code here
        
        for i in range(n):
            if nums[i] < 1 or nums[i] > n:
                nums[i] = 0
                
        for i in range(n):
            if 1 <= nums[i] % (n + 1) <= n:
                ind = nums[i] % (n + 1) - 1
                nums[ind] += n + 1
          
        for i in range(n):
            if nums[i] <= n:
                return i + 1
        
        return n + 1

#{ 
 # Driver Code Starts
#Initial Template for Python 3


import math


def main():
        T=int(input())
        while(T>0):
            
            n=int(input())
            
            arr=[int(x) for x in input().strip().split()]
            
            ob=Solution()
            print(ob.missingNumber(arr,n))
            
            T-=1


if __name__ == "__main__":
    main()
# } Driver Code Ends

File: Smallest Positive missing number - GFG/smallest-positive-missing-number.java
//{ Driver Code Starts
import java.util.*;


// } Driver Code Ends

class Solution
{
    //Function to find the smallest positive number missing from the array.
    static int missingNumber(int arr[], int size)
    {

    for(int i = 0; i < size; i++){
        int val = arr[i];
        if(val > 0 && val <= size){
            int idx = val -1;
            if(arr[idx] != val){
                int temp = arr[i];
                arr[i] = arr[idx];
                arr[idx] = temp;
                i--;                          ;//again we are checking the swapped element is at correct position
            }
        }
    }
    for(int i=0;i<size;i++){
     if(i+1!=arr[i])return i+1;

//if the element is not at its corect position then that will be the missing number (i+1)
 }

//if all are at its correct position then next element of the last element will be the missing number

//like 1st input test case [1,2,3,4,5] , next element i.e 6 will be the output
 return size+1;
    }
}


//{ Driver Code Starts.

class Main
{   
    public static void main (String[] args) 
    {

		Scanner sc=new Scanner(System.in);
		
        //taking testcases
		int t=sc.nextInt();
		while(t-->0){
		    
		    //input number n
			int n=sc.nextInt();
			int[] arr=new int[n];
			
			//adding elements to the array
			for(int i=0;i<n;i++)
				arr[i]=sc.nextInt();
				
			Solution obj = new Solution();
			
			//calling missingNumber()
			int missing = obj.missingNumber(arr,n);
			System.out.println(missing);
		}
    }
}


// } Driver Code Ends